test/inductor/test_distributed_patterns.py

```diff
@@ -499,6 +527,209 @@ def test_fake_distributed_inductor(self):
+                parallel: str,
+            ) -> torch.Tensor:
+                amax_a = a.abs().amax(dim=-1, keepdim=True)
+                amax_b_t = b_t.abs().amax(dim=-1, keepdim=True)
+                out = fp8_matmul(a, amax_a, b_t, amax_b_t, parallel)
+                ctx.save_for_backward(a, b_t, amax_b_t)
+                ctx.parallel = parallel
+                return out
+
+            @staticmethod
+            def backward(
+                ctx: torch.autograd.function.FunctionCtx, grad_out: torch.Tensor
+            ) -> tuple[torch.Tensor, torch.Tensor, None]:
+                a: torch.Tensor
+                b_t: torch.Tensor
+                amax_b_t: torch.Tensor
+                a, b_t, amax_b_t = ctx.saved_tensors
+                parallel = REVERSE[ctx.parallel]
+
+                # Workaround for https://github.com/pytorch/pytorch/issues/141881.
+                b_t = b_t + grad_out[0, :, None]
+
+                b = b_t.t().contiguous()
+                amax_grad_out = grad_out.abs().amax(dim=-1, keepdim=True)
+                amax_b = amax_b_t.t().amax(dim=-1, keepdim=True)
+                amax_b = amax_b.repeat_interleave(
+                    b.shape[0] // amax_b.shape[0], dim=0, output_size=b.shape[0]
+                )
+                grad_a = fp8_matmul(grad_out, amax_grad_out, b, amax_b, parallel)
+
+                tp_group = torch.distributed.group.WORLD
+                if parallel == "col":
+                    grad_out = funcol.all_gather_tensor(
+                        grad_out, gather_dim=0, group=tp_group
+                    )
+                if parallel == "row":
+                    a = funcol.all_gather_tensor(a, gather_dim=0, group=tp_group)
+                grad_b = grad_out.t() @ a
+
+                return grad_a, grad_b, None
+
+        @torch.compile(backend=aot_graph_capture_backend)
+        def ffn(x: torch.Tensor, w1: torch.Tensor, w2: torch.Tensor) -> torch.Tensor:
+            x = Fp8LinearFn.apply(x, w1, "col")
+            x = torch.nn.functional.relu(x)
+            x = Fp8LinearFn.apply(x, w2, "row")
+            return x
+
+        in_ = torch.randn(
+            (3072 // WORLD_SIZE, 4096), device="cuda", dtype=torch.bfloat16
+        )
+        w1 = torch.randn(
+            (8192 // WORLD_SIZE, 4096),
+            device="cuda",
+            dtype=torch.bfloat16,
+            requires_grad=True,
+        )
+        w2 = torch.randn(
+            (4096, 8192 // WORLD_SIZE),
+            device="cuda",
+            dtype=torch.bfloat16,
+            requires_grad=True,
+        )
+        ffn(in_, w1, w2)
+        # The main thing to look for here:
+        #  primals_1 is saved for backward.
+        #  in the backward graph, we all_gather(primals_1),
+        #  instead of saving its result for backward
+        self.assertExpectedInline(
+            str(fw_graph[0].code.strip()),
+            """\
+def forward(self, primals_1, primals_2, primals_3):
```

Comment:
hmm, is it safe to match such a large block of text?  I would be sort of surprised if we can rely on all of this exactly matching (e.g. small naming, ordering changes could happen)

1) wonder if for the fw graph, we just want to (a) identify 'first argument after self' by name, (b) verify that argument is returned from the graph, and then 
2) in the bw graph, we can add assertions that the same var as we returned from fw is being passed into all_gather?

Reply:
I mainly did it because it's easy to do with expecttests, and we have some precedent in the codebase. I agree it's a bit of overkill here though - I can probably just assert on e.g. the return string containing the `primal_1` output I want - happy to switch to that (and also that the backward contains the line where we allgather `primal_1` like you're pointing out.

Reply:
Ok. I'm mainly trying to avoid the test being over sensitive and causing false positives if some subtle thing changes in the graph. But if we are typically asserting based on graphs like this then this is probably fine.